# 4. Evaluation 
### 4.1 Processed Data without thresholds for Air-Quality
### 4.2 Processed Data with thresholds for Air-Quality

In [12]:
import pandas as pd
import pickle

# Load your dataset
df: pd.DataFrame = pd.read_csv('../data/AirQualityUCIClustered.csv')
with open('../data/features.pkl', 'rb') as file:
    features: list = pickle.load(file)

df_th: pd.DataFrame = pd.read_csv('../data/AirQualityUCIClusteredWithThresholds.csv')
with open('../data/features_with_thresholds.pkl', 'rb') as file:
    features_th: list = pickle.load(file)

In [13]:
# 4.1 Group by clusters and compute mean of features
cluster_summary = df.groupby('Cluster')[features].mean().round(2)
print("\n4.1 Cluster Summary:")
print(cluster_summary)

# Count of records per cluster
print("\nCluster Counts:")
print(df['Cluster'].value_counts())


4.1 Cluster Summary:
         CO(GT)  C6H6(GT)  NO2(GT)  PT08.S5(O3)  Month  HourOfDay      T  \
Cluster                                                                    
0          3.86     19.12   155.98      1541.78   6.56      14.10  16.32   
1          1.21      4.67    81.41       793.02   6.11       5.27  14.10   
2          1.82      9.34   104.92       905.46   6.33      16.19  24.09   

            RH  
Cluster         
0        53.91  
1        58.86  
2        35.31  

Cluster Counts:
Cluster
1    3544
2    3318
0    2495
Name: count, dtype: int64


In [14]:
# 4.2 Group by clusters and compute mean of features
cluster_summary_th = df_th.groupby('Cluster')[features_th].mean().round(2)
print("\n4.2 Cluster Summary (WHO included):")
print(cluster_summary_th)

# Count of records per cluster
print("\nCluster Counts:")
print(df_th['Cluster'].value_counts())


4.2 Cluster Summary (WHO included):
         CO(GT)  C6H6(GT)  NO2(GT)  PT08.S5(O3)  Month  HourOfDay      T  \
Cluster                                                                    
0          2.19     11.47   117.54      1113.71   6.30      13.25  20.15   
1          5.16     23.40   168.40      1674.10   7.48      14.62  17.23   
2          0.98      3.18    74.38       657.66   5.93       7.04  14.87   

            RH  CO(GT)ToHigh  C6H6(GT)ToHigh  NO2(GT)ToHigh  PT08.S5(O3)ToHigh  
Cluster                                                                         
0        45.81          0.00            1.00            1.0                1.0  
1        54.46          0.98            0.99            1.0                1.0  
2        53.96          0.01            0.07            1.0                1.0  

Cluster Counts:
Cluster
0    5532
2    2856
1     969
Name: count, dtype: int64


| Cluster | Count | Pollution Level    | Key Characteristics                                        | Time Pattern    | Environmental Conditions                                  | Interpretation                                                                                                                                   |
| ------- | ----- | ------------------ | ---------------------------------------------------------- | --------------- | --------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------ |
| **0**   | 5437  | Moderate Pollution | Moderate CO, high benzene, high NO₂, high O₃ sensor values | Around **1 PM** | Warmer temperature (~20°C) and lower humidity (~45%)      | Represents **typical daytime urban air quality**, likely influenced by normal traffic activity and photochemical reactions in the atmosphere.    |
| **1**   | 969   | High Pollution     | Highest CO, benzene, NO₂, and O₃ sensor values             | Around **3 PM** | Slightly cooler (~17°C) with higher humidity (~55%)       | Indicates **severe pollution episodes**, possibly caused by heavy traffic, industrial emissions, or atmospheric conditions that trap pollutants. |
| **2**   | 2951  | Low Pollution      | Lowest CO, benzene, NO₂, and O₃ sensor values              | Around **7 AM** | Coolest temperature (~15°C) with moderate humidity (~54%) | Represents **cleaner air conditions**, typically occurring in the early morning before peak traffic and daily activities increase emissions.     |


In [ ]:
# Adding Traffic-Light Colors: Green, Yellow and Red to dataset - representing the Air Quality in a simple way
df["AirQuality"] = df["Cluster"].map({0: "yellow", 1: "red", 2: "green"})

In [4]:
df.head()

,DateTime,Month,HourOfDay,CO(GT),PT08.S1(CO),C6H6(GT),PT08.S2(NMHC),NOx(GT),PT08.S3(NOx),NO2(GT),...,RH,AH,IsWeekend,IsWeekday,CO(GT)ToHigh,C6H6(GT)ToHigh,NO2(GT)ToHigh,PT08.S5(O3)ToHigh,Cluster,AirQuality
0,2004-03-10 18:00:00,3,18,2.6,1360.0,11.9,1046.0,166.0,1056.0,113.0,...,48.9,0.76,False,True,False,True,True,True,0,yellow
1,2004-03-10 19:00:00,3,19,2.0,1292.0,9.4,955.0,103.0,1174.0,92.0,...,47.7,0.73,False,True,False,True,True,True,0,yellow
2,2004-03-10 20:00:00,3,20,2.2,1402.0,9.0,939.0,131.0,1140.0,114.0,...,54.0,0.75,False,True,False,True,True,True,0,yellow
3,2004-03-10 21:00:00,3,21,2.2,1376.0,9.2,948.0,172.0,1092.0,122.0,...,60.0,0.79,False,True,False,True,True,True,0,yellow
4,2004-03-10 22:00:00,3,22,1.6,1272.0,6.5,836.0,131.0,1205.0,116.0,...,59.6,0.79,False,True,False,True,True,True,0,yellow


#### Quick Comparison of both Results (k-means vs DBSCAN)
Algorithm	    Result    
K-Means	        3 clusters (forced)     
DBSCAN	        2 clusters + noise     
Interpretation	Density structure suggests 2 main groups     